x/streamer/keeper/distribute.go

```diff
@@ -3,143 +3,147 @@ package keeper
 	// filling the epoch.
 	if stream.Sponsored {
 		distr, err := k.sk.GetDistribution(ctx)
 		if err != nil {
-			return sdk.Coins{}, fmt.Errorf("failed to get sponsorship distribution: %w", err)
+			return types.Stream{}, fmt.Errorf("get sponsorship distribution: %w", err)
 		}
-		info := types.DistrInfoFromDistribution(distr)
 		// Update stream distr info
-		stream.DistributeTo = info
+		stream.DistributeTo = types.DistrInfoFromDistribution(distr)
 	}
 
-	totalDistrCoins, err := k.DistributeByWeights(ctx, totalDistrCoins, stream.DistributeTo)
-	if err != nil {
-		return nil, err
-	}
+	// Add coins to distribute during the next epoch
+	stream.EpochCoins = epochCoins
 
-	err = k.updateStreamPostDistribute(ctx, stream, totalDistrCoins)
-	if err != nil {
-		return nil, err
+	// Don't fill streams in which there's nothing to fill. Note that rewards are distributed post factum.
+	// I.e., first increase the filled epoch number, then distribute rewards during the epoch.
+	if !stream.DistributeTo.TotalWeight.IsZero() {
+		stream.FilledEpochs += 1
 	}
 
-	ctx.EventManager().EmitEvents(sdk.Events{
-		sdk.NewEvent(
-			types.TypeEvtDistribution,
-			sdk.NewAttribute(types.AttributeStreamID, osmoutils.Uint64ToString(stream.Id)),
-			sdk.NewAttribute(types.AttributeAmount, totalDistrCoins.String()),
-		),
-	})
-	return totalDistrCoins, nil
+	return stream, nil
 }
 
-// updateStreamPostDistribute increments the stream's filled epochs field.
-// Also adds the coins that were just distributed to the stream's distributed coins field.
-func (k Keeper) updateStreamPostDistribute(ctx sdk.Context, stream types.Stream, newlyDistributedCoins sdk.Coins) error {
-	stream.FilledEpochs += 1
-	stream.DistributedCoins = stream.DistributedCoins.Add(newlyDistributedCoins...)
-	if err := k.setStream(ctx, &stream); err != nil {
-		return err
-	}
+type DistributeRewardsResult struct {
+	NewPointer       types.EpochPointer
+	FilledStreams    []types.Stream
+	DistributedCoins sdk.Coins
+	Iterations       uint64
+}
 
-	// Check if stream has completed its distribution
-	if stream.FilledEpochs >= stream.NumEpochsPaidOver {
-		if err := k.moveActiveStreamToFinishedStream(ctx, stream); err != nil {
-			return err
+// DistributeRewards distributes all streams rewards to the corresponding gauges starting with
+// the specified pointer and considering the limit.
+func (k Keeper) DistributeRewards(
+	ctx sdk.Context,
+	pointer types.EpochPointer,
+	limit uint64,
+	streams []types.Stream,
+) DistributeRewardsResult {
+	totalDistributed := sdk.NewCoins()
+
+	// Temporary map for convenient calculations
+	streamUpdates := make(map[uint64]sdk.Coins, len(streams))
+
+	// Distribute to all the remaining gauges that are left after EndBlock
+	newPointer, iterations := IterateEpochPointer(pointer, streams, limit, func(v StreamGauge) pagination.Stop {
+		distributed, errX := k.DistributeToGauge(ctx, v.Stream.EpochCoins, v.Gauge, v.Stream.DistributeTo.TotalWeight)
```

Comment:
don't it need to be in cache ctx?
semi-succefull `DistributeToGauge` should be reverted
otherwise the state can be some undesired state